File: server/flask_server/app/controller/stock_controller.py
import yfinance as yf
import pandas as pd
import os
current_directory = os.getcwd()


def fetch_stock_data(stock_name, start_date):
    try:
        data = yf.download(stock_name, start=start_date, progress=False)
        ticker_info = yf.Ticker(stock_name)
        long_name = ticker_info.info['longName']
        return data, long_name
    except ValueError:
        return None, None
    
def fetch_live_price(stock_name):
    try:
        ticker_info = yf.Ticker(stock_name)
        current_price = ticker_info.info['currentPrice']
        open_price = ticker_info.info['open']
        return current_price, open_price
    except ValueError:
        return None

# def fetch_all_stock_data():
#     current_file_directory = os.path.dirname(os.path.realpath(__file__))
#     file = os.path.join(current_file_directory, '..', 'datasets', 'nse.csv')
#     nse = pd.read_csv(file)
#     nse['SYMBOL'] = nse['SYMBOL'].apply(lambda x: x + '.NS')
#     nse = nse[nse['CLOSE'] > 10]
#     nse = nse[nse['SYMBOL'].str.contains(r'^[A-Z]+')]
#     nse = nse.reset_index(drop=True)
#     nse = nse.sort_values('SYMBOL')
#     nse = nse['SYMBOL']
#     return nse

def fetch_all_stock_data():
    current_file_directory = os.path.dirname(os.path.realpath(__file__))
    file = os.path.join(current_file_directory, '..', 'datasets', 'nse_stocks.csv')
    nse = pd.read_csv(file)
    nse = nse[['Ticker', 'Name']]
    nse = nse.sort_values('Ticker')
    nse = nse.reset_index(drop=True)
    return nse

File: flask_server_onnode/src/controller/stock_controller.js
const yahooFinance = require("yahoo-finance2");
const fs = require("fs");
const path = require("path");
const csvParser = require("csv-parser");

// Fetch historical stock data
const fetchStockData = async (stockName, startDate) => {
  try {
    const data = await yahooFinance.historical(stockName, { period1: startDate });
    const tickerInfo = await yahooFinance.quoteSummary(stockName, { modules: ["assetProfile"] });

    const longName = tickerInfo?.assetProfile?.longBusinessSummary || "Unknown";
    return { data, longName };
  } catch (error) {
    console.error(`Error fetching stock data: ${error.message}`);
    return { data: null, longName: null };
  }
};

// Fetch live stock price
const fetchLivePrice = async (stockName) => {
  try {
    const tickerInfo = await yahooFinance.quote(stockName);
    const currentPrice = tickerInfo?.regularMarketPrice || null;
    const openPrice = tickerInfo?.regularMarketOpen || null;

    return { currentPrice, openPrice };
  } catch (error) {
    console.error(`Error fetching live stock price: ${error.message}`);
    return { currentPrice: null, openPrice: null };
  }
};

// Fetch all stock data from CSV file
const fetchAllStockData = async () => {
  const filePath = path.join(__dirname, "..", "datasets", "nse_stocks.csv");
  const stocks = [];

  return new Promise((resolve, reject) => {
    fs.createReadStream(filePath)
      .pipe(csvParser())
      .on("data", (row) => stocks.push({ Ticker: row.Ticker, Name: row.Name }))
      .on("end", () => {
        stocks.sort((a, b) => a.Ticker.localeCompare(b.Ticker));
        resolve(stocks);
      })
      .on("error", (error) => {
        console.error(`Error reading CSV: ${error.message}`);
        reject([]);
      });
  });
};

module.exports = {fetchAllStockData,fetchLivePrice,fetchStockData}